qualcommbe: ipq95xx: add 8devices Kiwi-DVK support

mantas-p opened the pull request:
### Specifications
  SoC:     Qualcomm IPQ9570
  RAM:     1 GiB DDR4
  Flash:   4 GiB eMMC
  ETH1:    2.5G Maxlinear GPY211C (48V POE-in)
  ETH2:    10G Realtek RTL8261N
  SFP:     10G SFP+ cage
  WLAN1:   QCN9274 2.4GHz 802.11b/g/n/ax/be 2x2
  WLAN2:   QCN9274 5-6GHz 802.11a/n/ac/ax/be 2x2
  PCIE0:   Gen3x1 mPCIe or M.2(B+M)
  PCIE1:   Gen3x1 M.2(A+E) or M.2(B+M)
  PCIE2:   Gen3x2 M.2(A+E)
  USB:     1x 3.0
  Console: USB-C via FTDI
  Button:  Reset

### Install
Install using sysupgrade:
```
  sysupgrade openwrt-qualcommbe-ipq95xx-8devices_kiwi-dvk-squashfs-sysupgrade.bin
```
Install using U-Boot:

 * Put `openwrt-qualcommbe-ipq95xx-8devices_kiwi-dvk-squashfs-factory.bin` at TFTP server(192.168.2.254) as `fwupdate.bin`
 * Run `run do_recovery` in U-Boot shell

mantas-p committed:
qualcommbe: ipq95xx: add 8devices Kiwi-DVK support

Specifications:
  SoC:     Qualcomm IPQ9570
  RAM:     1 GiB DDR4
  Flash:   4 GiB eMMC
  ETH1:    2.5G Maxlinear GPY211C (48V POE-in)
  ETH2:    10G Realtek RTL8261N
  SFP:     10G SFP+ cage
  WLAN1:   QCN9274 2.4GHz 802.11b/g/n/ax/be 2x2
  WLAN2:   QCN9274 5-6GHz 802.11a/n/ac/ax/be 2x2
  PCIE0:   Gen3x1 mPCIe or M.2(B+M)
  PCIE1:   Gen3x1 M.2(A+E) or M.2(B+M)
  PCIE2:   Gen3x2 M.2(A+E)
  USB:     1x 3.0
  Console: USB-C via FTDI
  Button:  Reset

Install using sysupgrade:

  sysupgrade openwrt-qualcommbe-ipq95xx-8devices_kiwi-dvk-squashfs-sysupgrade.bin

Install using U-Boot:

 * Put factory.bin at TFTP server(192.168.2.254) as fwupdate.bin
 * Run 'run do_recovery' in U-Boot shell

Signed-off-by: Mantas Pucka <mantas@8devices.com>
Link: https://github.com/openwrt/openwrt/pull/18435
Signed-off-by: Robert Marko <robimarko@gmail.com>

aiamadeus commented on target/linux/qualcommbe/ipq95xx/config-default:
```diff
@@ -35,6 +40,8 @@ CONFIG_REGULATOR_QCOM_SPMI=y
 CONFIG_REGULATOR_USERSPACE_CONSUMER=y
 CONFIG_RTC_DRV_PM8XXX=y
+CONFIG_RTL8261N_PHY=y
+CONFIG_SFP=y
```
I think it would be better to use kmod-sfp in DEVICE_PACKAGES?

robimarko replied:
For me its fine to compile it in, this target wont be storage limited

aiamadeus replied:
But not all boards have sfp, bpi r3 / r4 also uses kernel module:
https://github.com/openwrt/openwrt/blob/66b5ed7a4eb5e8f9bc9fb1584b7fd97b0a9978bd/target/linux/mediatek/image/filogic.mk#L441-L448

Ansuel replied:
@robimarko I also think sfp as module is better... we can't use it for failsafe anyway so better make it ""appear"" after the rootfs is mounted.

mantas-p replied:
I don't mind either way, but shouldn't SFP be at least consistent with other PHY drivers (all compiled-in or all as modules)?

Ansuel replied:
@robimarko depends on the sfp module and the presence of it.

Ansuel commented:
massive work to finish the last bit... I plan to backport ath11k for the integrated ath11k wifi in the soc.

Ansuel commented:
I would like to cherry pick all the commit so that this gets reduced to a single one. Are you ok with this?

mantas-p commented:
> I would like to cherry pick all the commit so that this gets reduced to a single one. Are you ok with this?

Yes, of course, go ahead.

mantas-p commented:
> massive work to finish the last bit... I plan to backport ath11k for the integrated ath11k wifi in the soc.

Have you seen radio FW posted in the open? Their codelinaro repo only has BDF (https://git.codelinaro.org/clo/ath-firmware/ath11k-firmware/-/tree/main/IPQ9574/hw1.0?ref_type=heads). 

Btw this particular board doesn't use internal SoC radio so this is not a blocker.

Ansuel commented:
@mantas-p no but it must be somewhere... I feel if I make the work for the remote proc and the ath11k the fw might magically appear ;)

Doesn't use it in the sense that the rf plug are not populated? It's an optional feature of the soc?

mantas-p commented:
It's optional: IPQ9570 has no radio, only IPQ9574 has it.

Ansuel commented:
oh... guess defective silicon that they recycled as a cheaper soc...

aiamadeus commented on target/linux/qualcommbe/files/arch/arm64/boot/dts/qcom/ipq9570-kiwi-dvk.dts:
```diff
@@ -0,0 +1,460 @@
+		xgmac5: port@5 {
+			reg = <5>;
+			phy-mode = "2500base-x";
+			managed = "auto";
```
`managed = "auto";` is the default behavior, I think it's not needed?

aiamadeus commented on target/linux/qualcommbe/files/arch/arm64/boot/dts/qcom/ipq9570-kiwi-dvk.dts:
```diff
@@ -0,0 +1,460 @@
+	mmc-hs400-1_8v;
+	mmc-hs400-enhanced-strobe;
+	max-frequency = <384000000>;
+	bus-width = <8>;
```
Can these properties be sorted alphabetically?

aiamadeus commented on target/linux/qualcommbe/files/arch/arm64/boot/dts/qcom/ipq9570-kiwi-dvk.dts:
```diff
@@ -0,0 +1,460 @@
+&xo_clk {
+	clock-frequency = <48000000>;
+};
+
```
Extra blank lines.

robimarko commented on package/boot/uboot-tools/uboot-envtools/files/qualcommbe_ipq95xx:
```diff
@@ -0,0 +1,37 @@
+
+case "$board" in
+8devices,kiwi-dvk)
+	ubootenv_add_mmc "0:APPSBLENV" "0x0" "0x40000" "0x40000" "1"
```
No need for number of blocks, envtoools can autodetect those

robimarko commented on target/linux/qualcommbe/ipq95xx/config-default:
```diff
@@ -1,8 +1,11 @@
 CONFIG_DT_IDLE_GENPD=y
+CONFIG_F2FS_FS=y
 CONFIG_GRO_CELLS=y
+CONFIG_HWMON=y
```
If you need hwmon, use the existing kmod

mantas-p replied:
CONFIG_MAXLINEAR_GPHY (which is compiled-in) requires hwmon. Having hwmon be a module would force PHY driver to be a module as well.

mrnuke replied:
Why not make **phy-maxlinear** a kmod- package instead? See, for eaxmple `package/kernel/linux/modules/netdevices.mk`

mantas-p replied:
Initially I thought that just compiling-in would be a simpler approach (that's already used by mediatek/filogic), but I guess I'll try kmod now :)

mrnuke committed:
kernel: modules: support Maxlinear GPHY

Add support for Maxlinear GPHY module. Also add support for
kmod-polynomial, which is a dependency for this module.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18435
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke commented:
If you's like, feel free to give this patch a try for Maxlinear support:

```
From a3855961d3e3a0b94c71379b6b366219b55568ad Mon Sep 17 00:00:00 2001
From: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Date: Wed, 14 May 2025 23:51:59 -0500
Subject: [PATCH] kernel: modules: support Maxliner GPHY

Add support for Maxlinear GPHY module. Also add uspport for
kmod-polynomial, which is a dependency for this module.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
---
 package/kernel/linux/modules/hwmon.mk      |  9 +++++++++
 package/kernel/linux/modules/netdevices.mk | 15 +++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/package/kernel/linux/modules/hwmon.mk b/package/kernel/linux/modules/hwmon.mk
index 3e19a36cb0..a765976aac 100644
--- a/package/kernel/linux/modules/hwmon.mk
+++ b/package/kernel/linux/modules/hwmon.mk
@@ -746,3 +746,12 @@ endef
 $(eval $(call KernelPackage,hwmon-adcxx))
 
 
+define KernelPackage/polynomial
+  TITLE:=polynomial support
+  KCONFIG:=CONFIG_POLYNOMIAL
+  HIDDEN:=1
+  FILES:=$(LINUX_DIR)/lib/polynomial.ko
+  AUTOLOAD:=$(call AutoProbe, polynomial)
+endef
+
+$(eval $(call KernelPackage,polynomial))
diff --git a/package/kernel/linux/modules/netdevices.mk b/package/kernel/linux/modules/netdevices.mk
index ca0d33be78..bfe0b8a58b 100644
--- a/package/kernel/linux/modules/netdevices.mk
+++ b/package/kernel/linux/modules/netdevices.mk
@@ -255,6 +255,21 @@ endef
 
 $(eval $(call KernelPackage,et131x))
 
+define KernelPackage/phy-maxlinear
+   SUBMENU:=$(NETWORK_DEVICES_MENU)
+   TITLE:=Maxlinear Ethernet PHY driver
+   KCONFIG:=CONFIG_MAXLINEAR_GPHY
+   DEPENDS:=+kmod-libphy +kmod-hwmon-core +kmod-polynomial
+   FILES:=$(LINUX_DIR)/drivers/net/phy/mxl-gpy.ko
+   AUTOLOAD:=$(call AutoProbe,mxl-gpy)
+endef
+
+define KernelPackage/phy-maxlinear/description
+   Support Maxlinear Ethernet PHYs.
+endef
+
+$(eval $(call KernelPackage,phy-maxlinear))
+
 define KernelPackage/phy-microchip
    SUBMENU:=$(NETWORK_DEVICES_MENU)
    TITLE:=Microchip Ethernet PHY driver

```

mrnuke commented:
@mantas-p since #18796, v6.12 is available. Any chance you could give it a try?

mrnuke commented:
I tested **usxgmii** and **2500base-x**. When you say "don't work here", what are the symptoms? Do `wan`, `lan1`, and `lan2` fail to appear, or do they just not establish a link?

mantas-p commented:
@mrnuke I've got eth working, some PCS fixes were needed.

---

About using kmod-phy-maxlinear. Since qcom_ppe is compiled-in it takes whatever phy driver is available at the moment - which is Generic Clause 45 PHY and that is quite broken for maxlinear chip.

```
[    1.546040] qcom_ppe 3a000000.ethernet: EDMA configuration successful
[    1.548330] qcom_ppe 3a000000.ethernet lan1 (uninitialized): PHY [90000.mdio-1:01] driver [Generic Clause 45 PHY] (irq=POLL)
[    1.831765] qcom_ppe 3a000000.ethernet lan2 (uninitialized): PHY [90000.mdio-1:1c] driver [Realtek RTL8261N] (irq=POLL) 
```

Once kmod-phy-maxlinear is loaded it is not replaced as an active driver for phy. Of course you can do something like this:

```
echo 90000.mdio-1:01 > /sys/bus/mdio_bus/drivers/Generic\ Clause\ 45\ PHY/unbind
echo 90000.mdio-1:01 > /sys/bus/mdio_bus/drivers/Maxlinear\ Ethernet\ GPY211C/bind 
ifconfig lan1 down; ifconfig lan1 up
``` 

But I think it's not worth the complexity, and cost of having HWMON compiled in for this platform is basically zero anyway.

robimarko commented:
@mantas-p Just make the kmod load in preinit, so:
`AUTOLOAD:=$(call AutoProbe,mxl-gpy,1)`

That is how we use PHY drivers cause they need to be loaded early before ethernet is probed.

mantas-p commented:
@robimarko Is that supposed to work even for compiled-in ethernet drivers? Wouldn't compiled-in driver still get probed before pre-init runs? I tried both `AUTOLOAD:=$(call AutoProbe,mxl-gpy,1)` and `AUTOLOAD:=$(call AutoLoad,18,mxl-gpy,1)` with same result.

mantas-p commented:
Tried to pacakge `qcom-ppe` (and `pcs-qcom-ipq9574`) as kmods. This seems to fix load order. Would that be acceptable?

robimarko commented:
That is fine for me

mantas-p committed:
qualcommbe: package PPE and PCS drivers as kmods

This will allow more flexibility in using PHY drivers as kmods.

Signed-off-by: Mantas Pucka <mantas@8devices.com>
Link: https://github.com/openwrt/openwrt/pull/18435
Signed-off-by: Robert Marko <robimarko@gmail.com>

mantas-p commented on target/linux/qualcommbe/patches-6.12/0368-net-pcs-ipq-uniphy-allow-10gbase-r-use-with-phylink.patch:
```diff
@@ -0,0 +1,30 @@
+
+--- a/drivers/net/pcs/pcs-qcom-ipq9574.c
++++ b/drivers/net/pcs/pcs-qcom-ipq9574.c
+@@ -744,6 +744,7 @@ static int ipq_pcs_validate(struct phyli
+ 	switch (state->interface) {
+ 	case PHY_INTERFACE_MODE_SGMII:
+ 	case PHY_INTERFACE_MODE_QSGMII:
++	case PHY_INTERFACE_MODE_10GBASER:
+ 		return 0;
+ 	case PHY_INTERFACE_MODE_2500BASEX:
```
I see it working with at both 125000000 and 312500000 which is kind of suspicious that something is not getting applied and works only because of defaults.

mantas-p replied:
Well I see the change in /sys/kernel/debug/clk/clk_summary (nss_cc_port1_rx_clk for example) so I expected that these clocks to do something...

mrnuke committed:
qualcommbe: v6.12: update PCS patches to use .pcs_validate()

The original PCS driver did not use the .pcs_validate() in
phylink_pcs_ops. The patches for 1000/2500base-x and 10g modes were
taken from this old driver, and thus did not bother with
.pcs_validate(). Thus, even though these modes are supported, phylink
would not enable them.

To fix these, list these modes in .pcs_validate(). Also list them in
ipq_pcs_clk_rate_get(). While the latter does not appear to change
behavior, it does change the clock rates listed under
/sys/kernel/debug/clk

Co-developed-by: Mantas Pucka <mantas@8devices.com>
Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/18435
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke commented on target/linux/qualcommbe/patches-6.12/0361-net-pcs-Add-10GBASER-interface-mode-support-to-IPQ-U.patch:
```diff
@@ -1,18 +1,20 @@
 Date: Wed, 6 Mar 2024 17:40:52 +0800
-Subject: [PATCH] net: pcs: Add 10GBASER interface mode support to IPQ UNIPHY
- PCS driver
+Subject: [PATCH 1/8] net: pcs: Add 10GBASER interface mode support to IPQ
```
I suggest to use the `--no-numbered` argument to `git format-patch` when generating patches. This way, the title stays the same so there is one less diff to worry about.

mrnuke commented on package/boot/uboot-tools/uboot-envtools/files/qualcommbe_ipq95xx:
```diff
@@ -0,0 +1,37 @@
+		ubootenv_add_uci_config "/dev/mtd$idx" "${2}" "${3}" "${4}"
+}
+
+ubootenv_add_sys_mtd() {
+	local idx="$(find_mtd_index "${1}")"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_sys_config "/dev/mtd$idx" "${2}" "${3}" "${4}"
+}
```
Is `ubootenv_add_sys_mtd` used anywhere?

mrnuke reviewed (commented):
Any chance you could use https://github.com/mrnuke/openwrt/commit/23f639cd3d2486e9a407d16b6c0580f6343d20c1 for the first set of PCS fixups? There also some cases for 1000base-x and 10g_qxgmii that I discovered need to be addressed.

I also tested the PPE/PCS kmod patch, and I think the changes are all good.

mantas-p committed:
qualcommbe: add PCS driver fixes

After update to v6.12 some PCS driver fixes were dropped and additional
are needed due to updated driver.

Signed-off-by: Mantas Pucka <mantas@8devices.com>
Link: https://github.com/openwrt/openwrt/pull/18435
Signed-off-by: Robert Marko <robimarko@gmail.com>

mrnuke reviewed (approved):
From my perspective, this looks good.

xh0823 commented:
TP-LINKTL-7HDAP13000TCS-PoE/DC
Thank you for your efforts, and I hope that 10GPHY RTL8261N PHY2.5G QCA8082 can work soon!
I would also like to ask if the following hardware is already supported:
CPU IPQ 9570
Wireless chips QCN9274 * 3
2.5G PHY QCA8082
10G PHY RTL8261N

mantas-p commented:
> > @xh0823 HW for this PR has IPQ9570 + 1x QCN9274 + RTL8261N - so that works. No idea about the state of QCA8082.
> 
> Can you take a look at QCA8082? Unfortunately, there is no USB available. This is for disassembly and connection https://www.jb51.net/network/922573_all.html

Not really, I don't have this HW, nor time for it.

robimarko commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1911,6 +1926,16 @@ endef
+define KernelPackage/pcs-qcom-ipq9574
+  SUBMENU:=$(NETWORK_DEVICES_MENU)
+  TITLE:=Qualcomm IPQ9574 PCS driver
+  DEPENDS:=+kmod-phylink
```
Both of these should depend on qualcommbe target

robimarko commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1911,6 +1926,16 @@ endef
+  DEPENDS:=+kmod-phylink
+  KCONFIG:=CONFIG_PCS_QCOM_IPQ9574
+  FILES:=$(LINUX_DIR)/drivers/net/pcs/pcs-qcom-ipq9574.ko
+  AUTOLOAD:=$(call AutoLoad,20,pcs-qcom-ipq9574,1)
```
Same here, why autoload?

mantas-p replied:
Changed to AutoProbe

robimarko commented:
Thanks! Rebased on top of main and merged!
